contents/blog/how-we-turned-clickhouse-into-our-eventmansion.md

```diff
@@ -0,0 +1,84 @@
+
+## Turning ClickHouse into our EventMansion
+
+It wasn’t all smooth sailing. Mutations in particular tripped us up, as we initially used them a lot instead of taking the time to model the data perfectly for ClickHouse. The documentation suggests not running more than one or two mutations an hour. We were running several hundred per minute. 
+
+Our advice? Don’t do that. It will get you so far, but it’s not a sustainable approach and eventually lead to us experiencing an outage. 
+
+The reason for this is that, as you are mutating the data, ClickHouse essentially waits until it merge the mutated table parts in a single, larger table part. The more mutations you run, the further behind ClickHouse gets and the slower queries can get. This process is essentially very good for small numbers of large files, but not for large numbers of small files. It took us a long time to figure this out. 
+
+We also had some issues with ClickHouse’s indexes. If you know exactly what data you need to pull and can specify the exact key and row then that can end up slower than other requests because ClickHouse builds sparse indexes which aren’t a true index of the entire dataset. 
+
+However, beyond this the experience of migrating to ClickHouse was incredibly positive — especially if you compare to the way we solved our original priorities:
+
+- *Real-time results*: Our users want PostHog to stay as fast as it was in the beginning. ClickHouse’s performance improvements were such that, with a little cleverness in the way we architect, we’re still able to provide answers in real-time.
+- *Future-proofing:* We need to constantly grow our feature set. ClickHouse SQL was developed to resemble Postgres and MySQL syntax, which means we’ve been able to continue adding new features without ClickHouse becoming a bottleneck.
+- *Filter on user-defined properties:* We leverage ClickHouse’s JSON deserialization functionality and features such as [materialized columns](https://posthog.com/blog/clickhouse-materialized-columns) to provide this, albeit at the cost of some CPU overhead.
+
+## Making the switch
+
+Normally, the process of switching to a new database model could be quite tricky. Luckily however, we’d just created feature flags within PostHog and were able to leverage this when switching. 
+
+First, we deployed ClickHouse in parallel to Postgres and begane ingesting data simultaneously into both. We then put all of our features and internal tools behind a feature flag in PostHog, reimplemented our analytic queries one by one and then used the feature flags to switch over. 
+
+Once everything was migrated we were able to simply stop storing data on Postgres while our users began deploying PostHog through our Helm Chart, which we use because of the amazing [ClickHouse-Operator made by Altinity](https://altinity.com/kubernetes-operator/). Easy. 
```

Comment:
```suggestion
Once everything was migrated we were able to simply stop storing data on Postgres while our users began deploying PostHog through our [Helm Chart](https://github.com/PostHog/charts-clickhouse), which we use because of the amazing [ClickHouse-Operator made by Altinity](https://altinity.com/kubernetes-operator/). Easy. 
```